webui/react/src/hooks/useSettings.ts

```diff
@@ -0,0 +1,244 @@
+import queryString from 'query-string';
+import { useCallback, useEffect, useMemo, useState } from 'react';
+import { useHistory, useLocation } from 'react-router-dom';
+
+import { Primitive, RecordKey } from 'types';
+import { clone, isBoolean, isEqual, isNumber, isString } from 'utils/data';
+import { Storage } from 'utils/storage';
+
+import usePrevious from './usePrevious';
+import useStorage from './useStorage';
+
+export enum BaseType {
+  Boolean = 'Boolean',
+  Float = 'Float',
+  Integer = 'Integer',
+  String = 'String',
+}
+
+type GenericSettingsType = Primitive | Primitive[] | undefined;
+
+/*
+ * defaultValue     - Optional default value. `undefined` as ultimate default.
+ * skipUrlEncoding  - Avoid preserving setting in the URL query param.
+ * storageKey       - If provided, save/load setting into/from storage.
+ * type.baseType    - How to decode the string-based query param.
+ * type.isArray     - List based query params can be non-array.
+ */
+interface SettingsConfigProp {
+  defaultValue?: GenericSettingsType;
+  key: string;
+  skipUrlEncoding?: boolean;
+  storageKey?: string;
+  type: {
+    baseType: BaseType;
+    isArray?: boolean;
+  };
+}
+
+export interface SettingsConfig {
+  settings: SettingsConfigProp[];
+  storagePath: string;
+}
+
+type GenericSettings = Record<string, GenericSettingsType>;
+
+interface SettingsHook<T> {
+  resetSettings: (keys?: string[]) => void;
+  settings: T;
+  settingsCount: (keys?: string[]) => number;
```

Comment:
I could see a usage for this if this was to return the active key names (and not doing the counting), or if we exposed a separate hook that did this: if we wanted to compare two pages together eg for sharing webui state in sharable UI project. This can always be added later when we want to use it.

Reply:
great suggestion! generalized to `activeSettings` to provide more flexibility in usage.